### PR TITLE
Replace manual Tensor::cat KV cache with candle_nn::kv_cache::KvCache in SmolVLM text model

### LINKED ISSUE #903: [Feature]: Optimize autoregressive KV cache updates to avoid repeated tensor concatenation
### 🚀 Feature Description

Inside smolvlm the current autoregressive attention cache implementation grows the KV cache using repeated `Tensor::cat` operations during token generation.

Current implementation:

self.k_cache = Tensor::cat(&[&self.k_cache, &k], 1)?;
self.v_cache = Tensor::cat(&[&self.v_cache, &v], 1)?;

This reallocates and copies the entire cache on every decoding step.

For long-context autoregressive generation, this can lead to:
- repeated memory allocations
- increasing memory movement over time
- growing token latency as sequence length increases
- unnecessary bandwidth pressure during inference

This feature proposes replacing repeated concatenation with a preallocated KV cache and offset-based writes.

### 📂 Feature Category

Performance Optimization

### 💡 Motivation

Autoregressive transformer decoding is highly sensitive to KV cache efficiency.

The current implementation appends keys and values by concatenating tensors every generation step. Since the cache grows monotonically during decoding, this likely introduces O(n²) memory movement characteristics over long sequences.

Modern inference runtimes such as:
- llama.cpp
- vLLM
- TensorRT-LLM
- FlashInfer

typically avoid dynamic concatenation by using:
- preallocated KV cache buffers
- offset-based writes
- fixed-capacity cache management

Optimizing this path could improve:
- token generation latency
- long-context scalability
- memory efficiency
- throughput consistency during decoding

### 💭 Proposed Solution

Replace repeated KV cache concatenation with a preallocated cache strategy.

Potential approach:
- allocate KV cache tensors with a maximum context length
- maintain a current sequence offset/index
- write incoming K/V tensors into cache slices instead of reallocating
- update attention reads to operate on the active cache region only

Conceptually:

k_cache[:, current_pos:current_pos + seq_len, :] = k
v_cache[:, current_pos:current_pos + seq_len, :] = v

Additional considerations:
- configurable maximum context length
- cache reset behavior
- compatibility with future sliding-window attention support
- preserving current API behavior

### 📚 Library Reference

This optimization approach is commonly used in modern transformer inference runtimes including:
- llama.cpp
- vLLM
- TensorRT-LLM
- FlashInfer

It is also aligned with standard KV cache management approaches used in autoregressive LLM inference systems.

### 🔄 Alternatives Considered

Alternative optimizations considered:
- reducing dtype conversion overhead in attention
- BF16-only attention paths
- Flash Attention integration

However, avoiding repeated cache concatenation appears to be a more foundational optimization for long-context autoregressive decoding.

### 🎯 Use Cases

This would be especially valuable for:
- long-context text generation
- streaming multimodal conversations
- edge deployment scenarios with constrained memory bandwidth
- GPU inference workloads where allocation overhead becomes significant
- future low-latency inference optimization efforts

### 📝 Additional Context

_No response_

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Varun-sai-500**
## 📝 Description

**Fixes/Relates to:** #903 

Replaces the manual grow-by-concatenation KV cache in SmolText's attention layer with Candle's built-in KvCache.

The previous implementation appended new K/V tensors using:

Tensor::cat(&[&self.k_cache, &new_k], 1)

on every decode step. This reallocates and copies the entire accumulated cache each iteration, resulting in O(n²) total memory movement over a sequence of n tokens.

This PR migrates the attention cache to Candle's native KvCache, which preallocates storage up to max_seq_len and appends via indexed writes/slicing, reducing cache growth overhead to O(n) total over the sequence.

This primarily improves autoregressive decode efficiency and long-context scaling behavior.

---

## 🛠️ Changes Made

 Replaced k_cache: Tensor and v_cache: Tensor in Attention with a single KvCache
Replaced per-step Tensor::cat growth with self.cache.append(&k, &v)
Updated reset_cache() to use self.cache.reset() instead of reallocating empty tensors
Added candle_nn::kv_cache::KvCache integration
Centralized cache sizing using MAX_POSITION_EMBEDDINGS

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** (List new/updated tests) :
Criterion benchmarks (benches/kv_cache.rs) comparing manual cache concatenation vs KvCache over 1024 decode steps (NUM_HEADS=32, HEAD_DIM=64):

append_only_manual_cat_kv_cache_1024      2.785 s
append_only_candle_kvcache_1024           0.009 s
→ ~303x faster (isolated cache append cost)

attention_only_manual_cat_decode_1024     7.073 s
attention_only_candle_kvcache_decode_1024 4.691 s
→ ~1.5x faster (full attention decode path)

The append-only benchmark isolates pure cache growth/allocation overhead.

The attention benchmark reflects realistic decode conditions, where the quadratic attention matmul increasingly dominates runtime as sequence length grows, partially masking cache-management gains.

append_only_manual_cat_kv_cache_1024      2.785 s
append_only_candle_kvcache_1024           0.009 s    → ~303x faster (isolated cache cost)

attention_only_manual_cat_decode_1024     7.073 s
attention_only_candle_kvcache_decode_1024 4.691 s    → ~1.5x faster (full attention forward pass)

The append-only benchmark isolates pure allocation cost; the attention benchmark reflects real decode conditions where matmul compute partially masks the cache overhead at long sequences.

- [x] **Manual Verification:** (Describe the steps you took)
Ran full SmolVLM inference before and after — output logits match.

- [ ] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
Add any other context or screenshots about the pull request here.
he 303x figure reflects isolated allocation cost. In full attention forward passes the speedup is ~1.5x at 1024 steps because O(n²) matmul dominates at long sequences — the real fix for that is flash attention (existing TODO in the code). This PR addresses the cache allocation inefficiency specifically.

closes: #903

### CODE REVIEW

**Comment by sidd-27** [comment]
the second commit undoes the first one, just revert it @Varun-sai-500

File: crates/kornia-vlm/src/smolvlm/text_model.rs [hunk 1]
```diff
@@ -54,8 +53,7 @@ impl Attention {
         Ok(Self {
             cos: idx_theta.cos()?.to_dtype(q.dtype())?,
             sin: idx_theta.sin()?.to_dtype(q.dtype())?,
-            k_cache: Tensor::zeros((NUM_OF_HEADS, 0, HEAD_DIM), dtype, device)?,
-            v_cache: Tensor::zeros((NUM_OF_HEADS, 0, HEAD_DIM), dtype, device)?,
+            cache: KvCache::new(1, 16384),
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/text_model.rs]: this is a hardcoded value, why this? and what happens when this is exveeded, probaby a better way to handle this

Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs]: added a variable, at the top, so now it's handled, is it okay now?
It fails when exceeded

[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs]: Updated the code, now it has only one source of truth (max_position_embedding)

File: crates/kornia-vlm/src/smolvlm/text_model.rs [hunk 2]
```diff
@@ -107,30 +96,30 @@ impl Attention {
...
         let y = {
             // TODO: implement flash attention
             // TODO: just using BF16 is plausible
 
             let in_dtype = q.dtype();
             let q = q.to_dtype(DType::F32)?;
-            let k = self.k_cache.to_dtype(DType::F32)?;
-            let v = self.v_cache.to_dtype(DType::F32)?;
+            let k = k_cache.to_dtype(DType::F32)?;
+            let v = v_cache.to_dtype(DType::F32)?;
```
[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs]: Resolved now, now we don't do full cache conversion to FP32, we do it before the append

**Comment by Varun-sai-500** [comment]
### Benchmarks

Benchmarks were run on the CPU FP32 inference path using Criterion.

The previous implementation updated the KV cache by repeatedly rebuilding the
accumulated tensors using `Tensor::cat`, causing cache-update cost to grow with
context length during autoregressive decoding.

This change replaces the repeated concatenation path with `KvCache::append`.

### Isolated KV-cache update benchmark

| Cache Length | Manual `Tensor::cat` | `KvCache::append` | Speedup |
|---:|---:|---:|---:|
| 512 | 5.3016 ms | 4.1840 ms | 1.27× |
| 1024 | 10.275 ms | 5.2160 ms | 1.97× |
| 2048 | 18.174 ms | 6.7353 ms | 2.70× |
| 4096 | 38.120 ms | 9.5451 ms | 3.99× |
| 8192 | 78.188 ms | 17.476 ms | 4.47× |

At 8192 cached tokens, isolated KV-cache update latency improves from
78.188 ms to 17.476 ms, a 4.47× speedup.

### Synthetic attention decode-step benchmark

This benchmark includes:
- KV-cache update
- QKᵀ attention matmul
- softmax
- attention × V matmul

| Cache Length | Manual `Tensor::cat` | `KvCache` | Speedup |
|---:|---:|---:|---:|
| 1024 | 13.668 ms | 9.2866 ms | 1.47× |
| 2048 | 26.311 ms | 7.9371 ms | 3.31× |
| 4096 | 51.067 ms | 11.920 ms | 4.28× |
| 8192 | 82.837 ms | 40.139 ms | 2.06× |

At 4096 cached tokens, the synthetic decode-step benchmark improves from
51.067 ms to 11.920 ms, a 4.28× speedup.

### Synthetic autoregressive decode-loop benchmark

This benchmark simulates repeated token generation with long-context KV caches.

| Prefill Length | Decode Tokens | Manual `Tensor::cat` | `KvCache` | Speedup |
|---:|---:|---:|---:|---:|
| 2048 | 128 | 3.4859 s | 892.77 ms | 3.90× |
| 8192 | 128 | 14.147 s | 3.4549 s | 4.10× |

For an 8192-token prefill with 128 generated decode tokens, the synthetic
decode loop improves from 14.147 s to 3.4549 s, a 4.10× speedup.

The 16384-token append case is excluded from the headline comparison because
the benchmark initializes the cache at `MAX_POSITION_EMBEDDINGS` and then
appends one additional token, making it a boundary-overflow case rather than a
valid steady-state append benchmark.

File: crates/kornia-vlm/src/smolvlm/text_model.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs]: @sidd-27  should I push benchmark code as well? or just keep that in comments below?

**Comment by Varun-sai-500** [comment]
@sidd-27  kornia 3d issues are coming while testing this PR by running  pixi run rust-test  
What's the reason?

File: crates/kornia-vlm/src/smolvlm/text_model.rs [hunk 3]
```diff
@@ -107,35 +97,39 @@ impl Attention {
...
             let in_dtype = q.dtype();
+
             let q = q.to_dtype(DType::F32)?;
-            let k = self.k_cache.to_dtype(DType::F32)?;
-            let v = self.v_cache.to_dtype(DType::F32)?;
+            let k = k.to_dtype(DType::F32)?;
+            let v = v.to_dtype(DType::F32)?;
+
+            let (k_cache, v_cache) = self.cache.append(&k, &v)?;
+
```
[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs:119]: @edgarriba  worth a note, I always thought FP32 is used for precision issues, and also the TODO says "Using BF16 is plausible", so you'd want to implement that? @sidd-27

**Comment by edgarriba** [comment]
Hi @Varun-sai-500 — thanks for the contribution! The CI failures here aren't from your SmolVLM KV cache change. Your branch is 14 commits behind `main` (base `2bb7680`, main is at `48a074d`), and all the red checks are in files you didn't touch:

| Check | File |
|---|---|
| Rustfmt + Pre-commit | `kornia-py/src/pgo.rs:126` |
| TOML Format | `kornia-py/pyproject.toml` |
| Clippy | `crates/kornia-3d/src/ba_schur.rs:313,317,318` |
| Rust Test doctests | `crates/kornia-3d/src/lib.rs:13,20,38` (unicode math glyphs `×`, `−`, `⁻`, `¹` in doctest fences) |

These were fixed on `main` after you branched. A rebase should pick up the fixes and let CI focus on your actual change in `crates/kornia-vlm/src/smolvlm/text_model.rs`:

```bash
git fetch upstream
git rebase upstream/main
git push --force-with-lease
```

(Or `git merge upstream/main` if you prefer a merge.) Let us know once CI re-runs and we'll review.

**Comment by Varun-sai-500** [comment]
Oh sorry @edgarriba , I will fix this immediately

**Comment by Varun-sai-500** [comment]
@edgarriba  CI IS GREEN

**Review by edgarriba — APPROVED** [review]

File: crates/kornia-vlm/src/smolvlm/text_model.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/text_model.rs:119]: resolved in new PR
